auto_init_recursive_mutex() fails to compile in C++ 10

Using `g++` 10.3 against the `devel` or `master` branch of the SDK , the macro `auto_init_recursive_mutex(xxx)` generates a compile-time error:

````
/home/earle/Arduino/hardware/pico/rp2040//pico-sdk/src/common/pico_sync/include/pico/mutex.h:292:116: error: expected primary-expression before '.' token
  292 | #define auto_init_recursive_mutex(name) static __attribute__((section(".mutex_array"))) recursive_mutex_t name = { .core.spin_lock = (spin_lock_t *)1 /* marker for runtime_init */ }
      |                                                                                                                    ^
/home/earle/Arduino/sketch_jun20a/sketch_jun20a.ino:1:1: note: in expansion of macro 'auto_init_recursive_mutex'
    1 | auto_init_recursive_mutex(test);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/earle/Arduino/sketch_jun20a/sketch_jun20a.ino:1:27: warning: 'test' defined but not used [-Wunused-variable]
    1 | auto_init_recursive_mutex(test);
      |                           ^~~~
/home/earle/Arduino/hardware/pico/rp2040//pico-sdk/src/common/pico_sync/include/pico/mutex.h:292:107: note: in definition of macro 'auto_init_recursive_mutex'
  292 | #define auto_init_recursive_mutex(name) static __attribute__((section(".mutex_array"))) recursive_mutex_t name = { .core.spin_lock = (spin_lock_t *)1 /* marker for runtime_init */ }
      |                                                                                                           ^~~~
exit status 1
````

Compiling in `gcc` (i.e. straight C) succeeds without error, so this is a C++ vs. C issue.

My g++ command-line options (cut down to interesting bits):
````
arm-none-eabi-g++ -c -Wall -Wextra -Werror=return-type -Wno-ignored-qualifiers -march=armv6-m -mcpu=cortex-m0plus -mthumb \
    -iprefix/home/earle/Arduino/hardware/pico/rp2040/ @/home/earle/Arduino/hardware/pico/rp2040/lib/platform_inc.txt \
    -std=gnu++17 -g test.cpp -o test.o
````

I've also tried at `std=gnu++11` with the same error

--- a/src/common/pico_sync/include/pico/mutex.h
+++ b/src/common/pico_sync/include/pico/mutex.h
@@ -289,7 +289,7 @@ static inline bool recursive_mutex_is_initialized(recursive_mutex_t *mtx) {
  *
  * But the initialization of the mutex is performed automatically during runtime initialization
  */
-#define auto_init_recursive_mutex(name) static __attribute__((section(".mutex_array"))) recursive_mutex_t name = { .core.spin_lock = (spin_lock_t *)1 /* marker for runtime_init */ }
+#define auto_init_recursive_mutex(name) static __attribute__((section(".mutex_array"))) recursive_mutex_t name = { .core = { .spin_lock = (spin_lock_t *)1 /* marker for runtime_init */ }, .owner = 0, .enter_count = 0 }
 
 #ifdef __cplusplus
 }

--- a/test/kitchen_sink/kitchen_sink.c
+++ b/test/kitchen_sink/kitchen_sink.c
@@ -110,6 +110,9 @@ __force_inline int something_inlined(int x) {
     return x * 2;
 }
 
+auto_init_mutex(mutex);
+auto_init_recursive_mutex(recursive_mutex);
+
 int main(void) {
     spiggle();
 
@@ -118,6 +121,12 @@ int main(void) {
     printf("HI %d\n", something_inlined((int)time_us_32()));
     puts("Hello Everything!");
     puts("Hello Everything2!");
+
+    assert(mutex_try_enter(&mutex, NULL));
+    assert(!mutex_try_enter(&mutex, NULL));
+    assert(recursive_mutex_try_enter(&recursive_mutex, NULL));
+    assert(recursive_mutex_try_enter(&recursive_mutex, NULL));
     // this should compile as we are Cortex M0+
     __asm volatile("SVC #3");
+
 }

--- a/test/kitchen_sink/kitchen_sink.c
+++ b/test/kitchen_sink/kitchen_sink.c
@@ -122,10 +122,10 @@ int main(void) {
     puts("Hello Everything!");
     puts("Hello Everything2!");
 
-    assert(mutex_try_enter(&mutex, NULL));
-    assert(!mutex_try_enter(&mutex, NULL));
-    assert(recursive_mutex_try_enter(&recursive_mutex, NULL));
-    assert(recursive_mutex_try_enter(&recursive_mutex, NULL));
+    hard_assert(mutex_try_enter(&mutex, NULL));
+    hard_assert(!mutex_try_enter(&mutex, NULL));
+    hard_assert(recursive_mutex_try_enter(&recursive_mutex, NULL));
+    hard_assert(recursive_mutex_try_enter(&recursive_mutex, NULL));
     // this should compile as we are Cortex M0+
     __asm volatile("SVC #3");
 

